apps/emqx/src/emqx_persistent_session_ds/emqx_persistent_session_ds_stream_scheduler.erl

```diff
@@ -87,23 +111,41 @@ find_new_streams(S) ->
     %% after timeout?)
     Comm1 = emqx_persistent_session_ds_state:get_seqno(?committed(?QOS_1), S),
     Comm2 = emqx_persistent_session_ds_state:get_seqno(?committed(?QOS_2), S),
-    shuffle(
-        emqx_persistent_session_ds_state:fold_streams(
-            fun
-                (_Key, #srs{it_end = end_of_stream}, Acc) ->
-                    Acc;
-                (Key, Stream, Acc) ->
-                    case is_fully_acked(Comm1, Comm2, Stream) andalso not Stream#srs.unsubscribed of
-                        true ->
-                            [{Key, Stream} | Acc];
-                        false ->
-                            Acc
-                    end
-            end,
-            [],
-            S
-        )
-    ).
+    Filter = fun(_Key, Stream) -> is_fetchable(Comm1, Comm2, Stream) end,
+    #iter{
+        %% Limit the iteration to one round over all streams:
+        limit = emqx_persistent_session_ds_state:n_streams(S),
+        %% Filter out the streams not eligible for fetching:
+        filter = Filter,
+        %% Start the iteration right after the last visited stream:
+        it = emqx_persistent_session_ds_state:iter_streams(LastVisited, S),
+        %% Restart the iteration from the beginning:
+        it_cont = emqx_persistent_session_ds_state:iter_streams(beginning, S)
+    }.
+
+-spec next_stream(iter_stream()) -> {stream_key(), stream_state(), iter_stream()} | none.
+next_stream(#iter{limit = 0}) ->
+    none;
+next_stream(ItStream0 = #iter{limit = N, filter = Filter, it = It0, it_cont = ItCont}) ->
+    case emqx_persistent_session_ds_state:iterate(It0) of
+        {Key, Stream, It} ->
+            ItStream = ItStream0#iter{it = It, limit = N - 1},
+            case Filter(Key, Stream) of
+                true ->
+                    {Key, Stream, ItStream};
+                false ->
+                    next_stream(ItStream)
+            end;
+        none ->
+            %% Restart the iteration from the beginning:
+            ItStream = ItStream0#iter{it = ItCont},
+            next_stream(ItStream)
```

Comment:
Shouldn't we decrement `limit` before recursing here?

Reply:
Good point. We haven't gone through one more element, thus no decrement. The situation you're describing can happen in general, but should not here: `it_cont` returning `none` right away means that `limit` is `0`. But it still sounds like a good idea to defend from that honestly.